### PR TITLE
feat: support nested structs, structured arrays, and better secret value handling in config

### PR DESCRIPTION
**Pull request by f0ssel**
What this changes:
- Removes the need for `Path` field in config, we just use the json tags now
- Supports slices (does not currently support default values on structured slices)
- Support nested structs
- Supports structure array item field value overrides with `CODER_$INDEX_$FIELD`
- Safer secrets - by implementing `MarshalJSON` we can make sure we never return secrets fields over the wire.

### CODE REVIEW

File: codersdk/deploymentconfig.go [hunk 1]
```diff
@@ -12,77 +12,163 @@ import (
...
+			Name:       f.Name,
+			Usage:      f.Usage,
+			Flag:       f.Flag,
+			Shorthand:  f.Shorthand,
+			Enterprise: f.Enterprise,
+			Hidden:     f.Hidden,
+			Secret:     f.Secret,
+			Default:    f.Default,
+			Value:      f.Value,
+		})
```
Comment (kylecarbs) [inline codersdk/deploymentconfig.go:150]: Could we just marshal ourselves in this case?

```
json.Marshal(f)
```

**Review by kylecarbs — APPROVED** [review]
I like that this is structured similarly to the YAML... it's nice!

File: codersdk/deploymentconfig.go (hunk 1 above)
Comment (f0ssel) [reply codersdk/deploymentconfig.go:150]: You can't! I did that first and you get an infinite loop because it just calls this marshaled again. LOL

**Comment by f0ssel** [comment]
> I like that this is structured similarly to the YAML... it's nice!

It's exactly the structure, notice there are no mapstructure tags. The json and yaml will always stay in sync this way. I believe in the future we can implement a yaml marshaller and use it to write this struct to yaml, if for some reason we don't want to use viper for that.

**Pull request merged**